File: 프로그래머스/lv2/76502. 괄호 회전하기/괄호 회전하기.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
#include <stack>
using namespace std;

bool isMakeCorrect(string s){
    stack<char> st;
    
    int s_cnt = 0,m_cnt = 0,l_cnt = 0;
    for(auto c : s){
        if(c== '{' || c== '[' || c== '('){
            st.push(c);
        }
        else{
            if(s.size()==0)
                return false;
            if(c== '}'){
                if(st.top() == '{')
                    st.pop();
                else
                    return false;
            }
            if(c== ']'){
                if(st.top() == '[')
                    st.pop();
                else
                    return false;
            }
            if(c== ')'){
                if(st.top() == '(')
                    st.pop();
                else
                    return false;
            }
        }
    }
    if(st.size() == 0)
        return true;
    return false;
}

int solution(string s) {
    int answer = 0;
    for(int i=0;i<s.length();i++){
        rotate(s.begin(),s.begin()+1,s.end());
        if(isMakeCorrect(s))
            answer++;
    }
    return answer;
}

File: 프로그래머스/lv2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.Stack; 

class Solution {
    public int solution(String s) {
        Stack<Integer> stack = new Stack<Integer>();
        int answer = 0;
        for(int i=0;i<s.length();i++){
            String tmp = s.substring(1,s.length());
            s = tmp + s.charAt(0);
            stack.clear();
            boolean isMake = true;
            for(int j=0;j<s.length();j++){
                if(s.charAt(j) == '(')
                    stack.push(0);
                else if(s.charAt(j) == '{')
                    stack.push(1);
                else if(s.charAt(j) == '[')
                    stack.push(2);
                else{
                    if(stack.empty()){
                        isMake = false;
                        break;
                    } 
                    int popValue = -1;
                    if(s.charAt(j) == ')')
                        popValue = 0;
                    else if(s.charAt(j) == '}')
                        popValue = 1;
                    else
                        popValue = 2;
                    if(stack.peek() == popValue){
                        stack.pop();
                    }
                    else{
                        isMake = false;
                        break;
                    }
                }
            }
            if(isMake == true && stack.size()==0){
                answer++;                
            }
        }
        
        
        return answer;
    }
}